src/os/bluestore/BlueFS.cc

```diff
@@ -546,14 +549,25 @@ uint64_t BlueFS::get_block_device_size(unsigned id) const
   return 0;
 }
 
+void BlueFS::drop_pending_release(uint64_t delta)
+{
+  ceph_assert(delta <= pending_release);
+  pending_release -= delta;
+  if (wait_pending_release && pending_release == 0) {
+    pending_release_cond.notify_all();
+  }
+}
+
 void BlueFS::handle_discard(unsigned id, interval_set<uint64_t>& to_release)
 {
   dout(10) << __func__ << " bdev " << id << dendl;
   ceph_assert(alloc[id]);
+  ceph_assert(pending_release >= to_release.size());
   alloc[id]->release(to_release);
   if (is_shared_alloc(id)) {
     shared_alloc->bluefs_used -= to_release.size();
   }
+  drop_pending_release(to_release.size());
```

Comment:
Maybe this block should be common to `handle_discard` and inner part of `_release_pending_allocations` ?

Reply:
yep